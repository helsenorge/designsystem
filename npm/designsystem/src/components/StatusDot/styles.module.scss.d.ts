export type Styles = {
  statusdot: string;
  statusdot__dot: string;
  'statusdot__dot--active': string;
  'statusdot__dot--alert': string;
  'statusdot__dot--attachment': string;
  'statusdot__dot--cancelled': string;
  'statusdot__dot--draft': string;
  'statusdot__dot--exception': string;
  'statusdot__dot--group': string;
  'statusdot__dot--hidden': string;
  'statusdot__dot--inactive': string;
  'statusdot__dot--info': string;
  'statusdot__dot--inprocess': string;
  'statusdot__dot--inspected': string;
  'statusdot__dot--login': string;
  'statusdot__dot--noaccess': string;
  'statusdot__dot--on-dark': string;
  'statusdot__dot--pending': string;
  'statusdot__dot--recurring': string;
  'statusdot__dot--success': string;
  'statusdot__dot--transparent': string;
  'statusdot__dot--unknown': string;
  statusdot__label: string;
  'statusdot__label--on-dark': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
