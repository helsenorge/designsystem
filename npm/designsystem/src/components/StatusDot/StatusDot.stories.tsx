import React from 'react';

import { StoryObj, Meta } from '@storybook/react';

import StatusDot, { StatusDotVariant } from './StatusDot';
import Docs from '../../docs';

const meta = {
  title: '@helsenorge/designsystem-react/Components/StatusDot',
  component: StatusDot,
  parameters: {
    docs: {
      page: (): React.JSX.Element => <Docs component={StatusDot} />,
      description: {
        component: 'Viser en status i en kompakt og på en scanbar måte. Støtter en farget prikk og ikon.',
      },
    },
  },
  args: {
    variant: StatusDotVariant.info,
    text: 'StatusDot label',
  },
  argTypes: {
    variant: {
      control: 'select',
      options: StatusDotVariant,
    },
    text: {
      control: 'text',
    },
  },
} satisfies Meta<typeof StatusDot>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: args => <StatusDot {...args} />,
};
