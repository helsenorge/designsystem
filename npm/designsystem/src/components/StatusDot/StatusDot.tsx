import React from 'react';

import classNames from 'classnames';

import { IconSize } from '../..';
import { AnalyticsId } from '../../constants';
import { getColor } from '../../theme/currys';
import Attachment from '../Icons/Attachment';
import Group from '../Icons/Group';
import { Icon } from '../Icons/Icon';
import NoAccess from '../Icons/NoAccess';
import Undo from '../Icons/Undo';

import styles from './styles.module.scss';

export enum StatusDotModes {
  onwhite = 'onwhite',
  ondark = 'ondark',
}

export enum StatusDotVariant {
  info = 'info',
  warning = 'warning',
  alert = 'alert',
  cancelled = 'cancelled',
  active = 'active',
  transparent = 'transparent',
  recurring = 'recurring',
  group = 'group',
  noaccess = 'noaccess',
  attachment = 'attachment',
}

export interface StatusDotIconProps {
  /** The variant defines style formatting and what icon to use */
  variant?: keyof typeof StatusDotVariant;
}

const StatusDotIcon: React.FC<StatusDotIconProps> = ({ variant }) => {
  if (variant === StatusDotVariant.recurring) {
    return <Icon size={IconSize.XXSmall} svgIcon={Undo} />;
  } else if (variant === StatusDotVariant.group) {
    return <Icon size={IconSize.XXSmall} svgIcon={Group} />;
  } else if (variant === StatusDotVariant.noaccess) {
    return <Icon size={IconSize.XXSmall} svgIcon={NoAccess} color={getColor('cherry', 600)} />;
  } else if (variant === StatusDotVariant.attachment) {
    return <Icon size={IconSize.XXSmall} svgIcon={Attachment} />;
  }

  return null;
};

export interface StatusDotProps {
  /** id that is placed on the wrapper */
  id?: string;
  /** Defines the color mode, onwhite, ondark etc. */
  mode?: keyof typeof StatusDotModes;
  /** Visual variants for the statusdot */
  variant?: keyof typeof StatusDotVariant;
  /** Text placed to the right of the statusdot */
  text: string;
  /** Adds custom classes to the element. */
  className?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
}

const StatusDot: React.FC<StatusDotProps> = props => {
  const { id, mode = StatusDotModes.onwhite, variant = StatusDotVariant.info, text, className, testId } = props;

  const hasIcon =
    variant === StatusDotVariant.recurring ||
    variant === StatusDotVariant.group ||
    variant === StatusDotVariant.noaccess ||
    variant === StatusDotVariant.attachment;

  const isCancelled = variant === StatusDotVariant.cancelled;

  const statusDotClasses = classNames(styles['statusdot'], isCancelled && styles['statusdot--cancelled'], className);
  const dotClasses = classNames(styles['statusdot__dot'], hasIcon ? styles[`statusdot__dot--icon`] : styles[`statusdot__dot--${variant}`]);
  const labelClasses = classNames(
    styles['statusdot__label'],
    hasIcon && styles['statusdot__label--icon'],
    mode === StatusDotModes.ondark && styles['statusdot__label--on-dark']
  );

  return (
    <span id={id} className={statusDotClasses} data-testid={testId} data-analyticsid={AnalyticsId.StatusDot}>
      <span className={dotClasses}>
        <StatusDotIcon variant={variant} />
      </span>
      <span className={labelClasses}>{text}</span>
    </span>
  );
};

export default StatusDot;
