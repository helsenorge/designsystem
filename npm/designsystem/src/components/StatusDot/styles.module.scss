@use 'sass:map';
@use '../../scss/spacers' as spacers;
@use '../../scss/palette' as palette;
@use '../../scss/font-settings' as font-settings;
@use '../../scss/screen-reader' as *;
@use '../../scss/breakpoints' as breakpoints;
@import '../../scss/supernova/styles/colors.css';

$status-dot-size: 1rem;

.statusdot {
  position: relative;
  display: inline-flex;
  align-items: flex-start;
  font-size: $font-size-xs;
  line-height: 1.4rem;
  font-weight: 400;

  @media (min-width: map.get(breakpoints.$grid-breakpoints, sm)) {
    font-size: font-settings.$font-size-sm;
    line-height: 1.6rem;
  }

  &__dot {
    display: flex;
    align-items: center;
    justify-content: center;
    position: relative;
    top: 0.15rem;
    margin-right: spacers.getSpacer(2xs);
    width: $status-dot-size;
    height: $status-dot-size;
    min-width: $status-dot-size;
    min-height: $status-dot-size;

    @media (min-width: map.get(breakpoints.$grid-breakpoints, sm)) {
      top: spacers.getSpacer(3xs);
    }

    &--success {
      fill: var(--color-notification-graphics-success);
    }
    &--success#{&}--on-dark {
      fill: var(--core-color-kiwi-200);
    }

    &--inprocess,
    &--exception,
    &--unknown,
    &--inspected {
      fill: var(--color-notification-graphics-warning);
    }
    &--inprocess#{&}--on-dark,
    &--exception#{&}--on-dark,
    &--unknown#{&}--on-dark,
    &--inspected#{&}--on-dark {
      fill: var(--core-color-banana-300);
    }

    &--cancelled,
    &--alert {
      fill: var(--color-notification-graphics-error);
    }
    &--cancelled#{&}--on-dark,
    &--alert#{&}--on-dark {
      fill: var(--core-color-cherry-100);
    }

    &--info {
      fill: var(--core-color-blueberry-500);
    }
    &--info#{&}--on-dark {
      fill: var(--core-color-blueberry-200);
    }

    &--transparent {
      fill: var(--color-disabled-border);
    }
    &--transparent#{&}--on-dark,
    &--group#{&}--on-dark,
    &--recurring#{&}--on-dark,
    &--noaccess#{&}--on-dark,
    &--draft#{&}--on-dark,
    &--hidden#{&}--on-dark,
    &--login#{&}--on-dark,
    &--attachment#{&}--on-dark {
      fill: var(--color-base-graphics-ondark);
    }

    svg {
      position: absolute;
    }
  }

  &__label {
    &--on-dark {
      color: $white;
    }
  }
}
