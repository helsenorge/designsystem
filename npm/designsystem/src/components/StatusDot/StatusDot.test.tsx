import * as React from 'react';

import { render, screen } from '@testing-library/react';

import StatusDot from '.';

describe('Gitt at StatusDot rendres ', () => {
  describe('Når komponenten får en tekst', () => {
    it('Så skal teksten vises til innbygger', () => {
      render(<StatusDot variant="warning" text="Eksempeltekst" />);

      const text = screen.getByText('Eksempeltekst');
      expect(text).toBeVisible();
    });
  });
  describe('Når StatusDot er variant="noaccess".', () => {
    it('Så skal komponenten ha en SVG-fil.', () => {
      const { container } = render(<StatusDot variant="noaccess" text="Eksempeltekst" testId={'statusdot'} />);

      const text = screen.getByText('Eksempeltekst');
      expect(text).toBeVisible();

      expect(container).toMatchSnapshot();
    });
  });
  describe('Når StatusDot er en av de andre variantene', () => {
    it('Så skal komponenten ikke ha en tilhørende SVG-fil og samsvarende klassenavn', () => {
      render(
        <>
          <StatusDot variant="info" text="text1" testId={'statusdot1'} />
          <StatusDot variant="warning" text="text2" testId={'statusdot2'} />
          <StatusDot variant="alert" text="text3" testId={'statusdot3'} />
          <StatusDot variant="transparent" text="text4" testId={'statusdot4'} />
        </>
      );

      const dot1 = screen.getByTestId('statusdot1').firstElementChild;
      const dot2 = screen.getByTestId('statusdot2').firstElementChild;
      const dot3 = screen.getByTestId('statusdot3').firstElementChild;
      const dot4 = screen.getByTestId('statusdot4').firstElementChild;

      expect(dot1.className).toBe('statusdot__dot statusdot__dot--info');
      expect(dot2.className).toBe('statusdot__dot statusdot__dot--warning');
      expect(dot3.className).toBe('statusdot__dot statusdot__dot--alert');
      expect(dot4.className).toBe('statusdot__dot statusdot__dot--transparent');
    });
  });
});
