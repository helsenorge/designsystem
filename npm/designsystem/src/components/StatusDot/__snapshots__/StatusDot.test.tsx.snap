// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at StatusDot rendres  Når StatusDot er variant="attachment". Så skal komponenten ha en SVG-fil. 1`] = `
<div>
  <span
    class="statusdot"
    data-analyticsid="status-dot"
    data-testid="statusdot"
  >
    <span
      class="statusdot__dot statusdot__dot--icon"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        data-analyticsid="icon"
        fill="black"
        focusable="false"
        height="24"
        role="presentation"
        style="min-width: 24px; min-height: 24px;"
        viewBox="0 0 48 48"
        width="24"
      >
        <path
          d="M24 9.172c-3.597 0-6.542 2.768-6.542 6.199v22.46h2.01v-22.46c0-2.295 2.019-4.189 4.531-4.189 2.514 0 4.53 1.894 4.53 4.189v19.827c0 1-.709 1.62-1.735 1.62-1.026 0-1.723-.682-1.723-1.62V18.934H23.06v16.264c0 2.282 1.668 3.63 3.735 3.63 2.068 0 3.748-1.661 3.748-3.63V15.371c0-3.43-2.945-6.2-6.543-6.2Z"
        />
      </svg>
    </span>
    <span
      class="statusdot__label statusdot__label--icon"
    >
      Eksempeltekst
    </span>
  </span>
</div>
`;

exports[`Gitt at StatusDot rendres  Når StatusDot er variant="group". Så skal komponenten ha en SVG-fil. 1`] = `
<div>
  <span
    class="statusdot"
    data-analyticsid="status-dot"
    data-testid="statusdot"
  >
    <span
      class="statusdot__dot statusdot__dot--icon"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        data-analyticsid="icon"
        fill="black"
        focusable="false"
        height="24"
        role="presentation"
        style="min-width: 24px; min-height: 24px;"
        viewBox="0 0 48 48"
        width="24"
      >
        <path
          d="M8.812 5.18h-.274a4.639 4.639 0 0 0-4.644 4.65v2.276a4.639 4.639 0 0 0 4.644 4.65h.274a4.64 4.64 0 0 0 4.644-4.65V9.83a4.64 4.64 0 0 0-4.644-4.65m4.07 14.526c-.076-.05-.15-.076-.226-.126a8.236 8.236 0 0 1-3.844.926h-.274a8.113 8.113 0 0 1-3.796-.926A8.612 8.612 0 0 0 0 27.306v2.1h17.376v-2.1c0-3.3-1.798-6.15-4.494-7.6m12.356-5.55a8.235 8.235 0 0 1-3.844.924h-.274a8.104 8.104 0 0 1-3.794-.924 9.173 9.173 0 0 0-2.772 2.198c4.044 2.152 6.566 6.3 6.566 10.952v2.1h8.838V21.88c0-3.4-1.898-6.3-4.72-7.724m.864-9.504v2.282a4.646 4.646 0 0 1-4.644 4.65h-.274a4.648 4.648 0 0 1-4.646-4.65V4.652c0-2.57 2.08-4.652 4.646-4.652h.274a4.647 4.647 0 0 1 4.644 4.652"
          fill-rule="evenodd"
          transform="translate(8, 8.2966)"
        />
      </svg>
    </span>
    <span
      class="statusdot__label statusdot__label--icon"
    >
      Eksempeltekst
    </span>
  </span>
</div>
`;

exports[`Gitt at StatusDot rendres  Når StatusDot er variant="noaccess". Så skal komponenten ha en SVG-fil. 1`] = `
<div>
  <span
    class="statusdot"
    data-analyticsid="status-dot"
    data-testid="statusdot"
  >
    <span
      class="statusdot__dot statusdot__dot--icon"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        data-analyticsid="icon"
        fill="#B62E1C"
        focusable="false"
        height="24"
        role="presentation"
        style="min-width: 24px; min-height: 24px;"
        viewBox="0 0 48 48"
        width="24"
      >
        <path
          d="M13.4 26.1h21.2v-4.2H13.4v4.2ZM24 9C15.728 9 9 15.728 9 24s6.728 15 15 15c8.27 0 15-6.728 15-15S32.27 9 24 9Z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="statusdot__label statusdot__label--icon"
    >
      Eksempeltekst
    </span>
  </span>
</div>
`;

exports[`Gitt at StatusDot rendres  Når StatusDot er variant="recurring". Så skal komponenten ha en SVG-fil. 1`] = `
<div>
  <span
    class="statusdot"
    data-analyticsid="status-dot"
    data-testid="statusdot"
  >
    <span
      class="statusdot__dot statusdot__dot--icon"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        data-analyticsid="icon"
        fill="black"
        focusable="false"
        height="24"
        role="presentation"
        style="min-width: 24px; min-height: 24px;"
        viewBox="0 0 48 48"
        width="24"
      >
        <path
          d="M38.726 24.194c0 8.196-6.668 14.864-14.864 14.864-7.028 0-13.152-4.992-14.56-11.87l2.744-.56c1.14 5.578 6.11 9.628 11.816 9.628 6.652 0 12.064-5.41 12.064-12.062 0-6.652-5.412-12.062-12.064-12.062-3.1 0-6.05 1.218-8.272 3.312h5.248v2.802h-9.516l-.004-9.89h2.802v4.624a14.893 14.893 0 0 1 9.742-3.648c8.196 0 14.864 6.666 14.864 14.862"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="statusdot__label statusdot__label--icon"
    >
      Eksempeltekst
    </span>
  </span>
</div>
`;
